types/logger/logger.go

```diff
@@ -92,10 +97,20 @@ func RateLimitedFn(logf Logf, f time.Duration, burst int, maxCache int) Logf {
 		block
 	)
 
-	judge := func(format string) verdict {
+	// judge decides the fate of a log request and returns the string that should be used
+	// to describe the format when the verdict is warn.
+	judge := func(format string, args ...interface{}) (v verdict, warnFormat string) {
+		for _, arg := range args {
+			switch arg := arg.(type) {
+			case noRateLimit:
+				return allow, ""
+			}
+		}
+		format = noopFormatRemover.Replace(format)
```

Comment:
can we relay this call until it's needed on line 139? I don't think the line 112 cares.

Reply:
It's needed on line 119, but delayed past the `rateFreePrefix` loop.

Reply:
why do we need it on line 119? it's fine if there's a %5.2L gibberish in there, as users don't see it.

I'd rather not allocate on each rate-limited log print.

Reply:
Moved. Did it in a follow-up, because the rebase got ugly.